library/swift/src/GRPC/GRPCResponseHandler.swift

```diff
@@ -0,0 +1,175 @@
+    override init() {
+        super.init()
+        self.underlyingHandler
+            .onHeaders { [weak self] headers, _, endStream in
+                self?.handleHeaders(headers)
+                if endStream {
+                    self?.handleCompletion(trailers: [:])
+                }
+            }
+            .onData { [weak self] data, endStream in
+                self?.handleData(data)
+                if endStream {
+                    self?.handleCompletion(trailers: [:])
+                }
+            }
+            .onTrailers { [weak self] trailers in
+                self?.handleCompletion(trailers: trailers)
+            }
+            .onError { [weak self] _ in
+                // TODO: Switch to call a different handler with the error
+                self?.handleCompletion(trailers: [:])
+            }
+    }
+
+    // MARK: - Private
+
+    private func handleHeaders(_ headers: [String: [String]]) {
+        let grpcStatus = headers["grpc-status"]?
+            .first
+            .flatMap(Int.init)
+        self.onHeaders?(headers, grpcStatus)
+    }
+
+    private func handleData(_ data: Data) {
+        if data.isEmpty {
+            return
+        }
+
+        if var existingBuffer = self.buffer {
+            existingBuffer.append(contentsOf: data)
+        } else {
+            self.buffer = data
+        }
+
+        self.processBuffer()
+    }
+
+    private func handleCompletion(trailers: [String: [String]]) {
+        self.onComplete?(trailers)
+    }
+
+    private func processBuffer() {
+        assert(self.buffer != nil, "Buffer should never be nil when processing")
+
+        switch self.state {
+        case .expectingCompressionFlag:
+            guard let compressionFlag: UInt8 = self.buffer.integer(at: 0) else {
+                return
+            }
+
+            guard compressionFlag == 0 else {
+                assertionFailure("gRPC decompression is not supported")
+                self.buffer.removeAll()
+                self.state = .expectingCompressionFlag
+                return
+            }
+
+            self.buffer.removeFirst()
+            self.state = .expectingMessageLength
+
+        case .expectingMessageLength:
+            guard let messageLength: UInt32 = self.buffer.integer(at: 0) else {
+                return
+            }
+
+            self.buffer.removeFirst(MemoryLayout<UInt32>.size)
+            self.state = .expectingMessage(length: messageLength)
+
+        case .expectingMessage(let length):
+            let lengthAsInt = Int(length)
+            self.onMessage?(self.buffer.withUnsafeBytes { rawBufferPointer -> Data in
+                return Data(bytes: rawBufferPointer.baseAddress!, count: lengthAsInt)
+            })
+
+            self.buffer.removeFirst(lengthAsInt)
+            self.state = .expectingCompressionFlag
+        }
+
+        self.processBuffer()
+    }
+}
+
+private extension Data {
+    func integer<T: FixedWidthInteger>(at index: Int) -> T? {
```

Comment:
Not sure how "final" this is, but there are better ways of doing this unsafe cast.

Reply:
Updated, let me know if you had something else in mind 😃